### PR TITLE
Improvements to Fuzzy Control System

### PR DESCRIPTION
**Pull request by jsexauer**
@JDWarner  Building on #70

This is still a work-in-progress.  I just wanted to get your opinion before going too far down it to make sure you like the direction.

Major changes:
- Refactor code to use a `FuzzyVariableAdjectivie` object.  This object is what actually holds the membership function istead of the FuzzyVariable.  So in your version `fuzzyvar['poor']` was the actual membership function while in my version `fuzzyvar['porr']` is an object.  This has a lot of benefits that I hope you will see improve the flow of the code.
- Refactor the rules and control objects to work directly off the graph.  It seemed strange to me you had in essence parallel data structures going: a graph to control the firing and a set of dictionaries to track other flow.  So I consolidated this down to a single data structure, the graph.
- Preliminary support for complements of fuzzy adjectives
- Prelimimary support for intermediary fuzzy variables (ie, antecedent -> rule -> intermediary -> rule -> consequence)
- I removed your implementation of caching but plan to add that all back in later on.  I was planning to do it at the individual objects levels instead of trying to make the controller track that.  I think it will be a lot cleaner.

Let me know what you think.

### CODE REVIEW

File: skfuzzy/control/fuzzyvariable.py [hunk 1]
```diff
@@ -31,7 +69,7 @@ class FuzzyVariable(object):
     This class is designed as the base class underlying the Antecedent and
     Consequent classes, not for individual use.
     """
-    def __init__(self, universe, label):
+    def __init__(self, universe, label, defuzzy_method='centroid'):
```
Comment (JDWarner) [inline skfuzzy/control/fuzzyvariable.py]: `defuzz_method` or `defuzzify_method` would be more consistent with existing names in the package. Though if there is a strong case for 'defuzzy' we could deprecate those old names.

File: skfuzzy/control/fuzzyvariable.py [hunk 2]
```diff
@@ -44,12 +82,30 @@ def __init__(self, universe, label):
...
         self.label = label
-        self.connections = OrderedDict()
+        self.defuzzy_method = defuzzy_method
+        self.adjectives = OrderedDict()
+
         self._id = id(self)
-        self.output = None
+        self._crisp_value_accessed = False
+
+        class _NotGenerator(object):
```
Comment (JDWarner) [inline skfuzzy/control/fuzzyvariable.py:92]: Nice approach! Should we force users to conform to exactly 'NOT-' as a prefix, or consider matching the first 4 characters against a real expression that doesn't care about case, and accepts whitespace or a dash as a separator? I'd prefer to type `variable['not mylabel']` over `variable['NOT-mylabel']`. Perhaps a little magic would be acceptable here.

File: skfuzzy/control/fuzzyvariable.py [hunk 3]
```diff
@@ -84,76 +139,106 @@ def __getitem__(self, key):
...
+                             "crisp value of this variable.")
+
+        mf = item.mf
 
         if mf.size != self.universe.size:
             raise ValueError("New membership function {0} must be equivalent "
                              "in length to the universe variable.\n"
                              "Expected {1}, got {2}.".format(
-                                 key, self.universe.size, item.size))
+                                 key, self.universe.size, mf.size))
```
Comment (JDWarner) [inline skfuzzy/control/fuzzyvariable.py:168]: Good catch

File: skfuzzy/control/fuzzyvariable.py [hunk 4]
```diff
@@ -11,6 +14,41 @@
     from .ordereddict import OrderedDict
 
 
+class FuzzyVariableAdjective(object):
```
Comment (JDWarner) [inline skfuzzy/control/fuzzyvariable.py]: If I'm following, this class is intended to replace the (admittedly less than ideal) `self.active` structure for tracking the membership function of interest. I like the approach, but wonder if `Adjective` is the best name. It implies modification, where this doesn't really change anything - it just highlights a single membership function.

Something like "Active", "Operative," or "Emphasis" might be a bit more intuitively named. I'd lean toward "FuzzyVariableOperative"

**Comment by JDWarner** [comment]
@jsexauer Thanks for this, I like what you've done with the place! Lots of cleanup, noticed several bug fixes, and I like your direction.

I made a few minor comments but haven't yet had the time to really dive in. Tomorrow hopefully I'll have more time to kick the tires. It would be nice to have the caching back, but cleaning up the overall design is most important.

Some of the tests will need to be changed with the new design, but I'm not really concerned about this.

Feel free to add your name to CONTRIBUTORS.txt!

File: skfuzzy/control/fuzzyvariable.py (hunk 2 above)
Comment (jsexauer) [inline skfuzzy/control/fuzzyvariable.py:92]: users should not be using variable['NOT-adjective'], they should use variable.not_['adjective'].

File: skfuzzy/control/fuzzyvariable.py (hunk 4 above)
Comment (jsexauer) [inline skfuzzy/control/fuzzyvariable.py]: This does much more than replace the `self.active` structure.  It's trying to mimic how fuzzy variables are defined in IEC 61131-7 (the fuzzy control language).  They implement a variable like this:

```
    FUZZIFY Variable1
        TERM Bad := (0, 0) (0, 1) (50, 0) ; (* aka the adjective Variable1['Bad'] *)
        TERM Normal := (14, 0) (50, 1) (83, 0) ;
        TERM Good := (50, 0) (100, 1) (100, 0) ;
    END_FUZZIFY
```

Regarding the name adjective, I stole it from pyfuzzy's implementation, as that is what they call a "TERM."  To me it makes a lot of sense, but I guess we can change it.

http://pyfuzzy.sourceforge.net/doc/class_diagram.png

File: skfuzzy/control/fuzzyvariable.py (hunk 1 above)
Comment (jsexauer) [inline skfuzzy/control/fuzzyvariable.py]: Yea no problem.

**Comment by jsexauer** [comment]
Thanks!  Here's a mini example that shows off how the new API works:

``` python
# Inputs: qualtiy, service, decor
# Outpus: Tip
# Intermediary: ambiance
decor = fuzz.Antecedent(np.arange(0, 11, 1), 'decor')
decor.automf(3)

ambiance = fuzz.Intermediary(np.arange(0, 11, 1), 'ambiance')
ambiance.automf(3)

# If service is poor and decor is not good, ambiance is poor
rule4 = fuzz.Rule([service['poor'], decor.not_['good']], ambiance['poor'])
rule2.view()

# If ambiance is poor, tip is poor
rule5 = fuzz.Rule(ambiance['poor'], tip['poor'])
```

File: skfuzzy/control/fuzzyvariable.py (hunk 4 above)
Comment (JDWarner) [inline skfuzzy/control/fuzzyvariable.py]: In that case, let's go straight for the FCL standard and use "Term". In the long run that's probably the most intuitive.

**Comment by JDWarner** [comment]
@jsexauer would you like me to merge this as well as skfuzzy/skfuzzy#70 so you can work off of the master repository? Seems like that might be cleaner, and it would get more public exposure. 

I'd merge this with failing test(s), then raise an issue to fix the tests. Right now ironing out the API is more important than 100% coverage.

**Comment by jsexauer** [comment]
Sounds good to me.

**Comment by JDWarner** [comment]
All right, thanks @jsexauer! Merging.

**Pull request merged**